### PR TITLE
[4.2] Allow to disable session metadata tracking for guest users

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes

This should improve performance for sites that does not use User registration/login feature.

We already have an option to disable "Track Session Metadata",
However it more usefull to be able to disable it only for non registered users, while keep track of metadata for loged in users.
That what I made here.

### Testing Instructions

Enable debug with DB queries enabled. 
Visit the site frontend (do not login) and check how much DB queries.

Apply patch.
Go to global configuration and set "Track Session Metadata for non registered Users" to "No".
Visit the site frontend again (do not login) and check how much DB queries.

### Expected result AFTER applying this Pull Request
All should continue work as before, without any errors.

And it should reduce 2db query.

### Documentation Changes Required

Yes, I guess, need a description for this feature.

### CODE REVIEW

**Comment by dgrammatiko** [comment]
Does it also drops the cookie for non logged in users?

**Comment by Fedik** [comment]
> Does it also drops the cookie for non logged in users?

hm, nope, session still works as before, it just does not store "User  state" in database.

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on a1e36cf4798134aae0770fc34703e4787a4d9127<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37459">issues.joomla.org/tracker/joomla-cms/37459</a>.</sub>

File: administrator/components/com_config/forms/application.xml [hunk 1]
```diff
@@ -951,6 +951,20 @@
...
+		<field
+			name="session_metadata_guest"
+			type="radio"
+			layout="joomla.form.field.radio.switcher"
+			default="0"
+			label="COM_CONFIG_FIELD_SESSION_METADATA_GUEST_LABEL"
+			description="COM_CONFIG_FIELD_SESSION_METADATA_GUEST_DESC"
+			filter="boolean"
+			showon="session_metadata:1"
+		>
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			>
```

File: administrator/components/com_config/forms/application.xml [hunk 2]
```diff
@@ -951,6 +951,20 @@
 			<option value="1">JYES</option>
 		</field>
 
+		<field
+			name="session_metadata_guest"
+			type="radio"
+			layout="joomla.form.field.radio.switcher"
+			default="0"
+			label="COM_CONFIG_FIELD_SESSION_METADATA_GUEST_LABEL"
+			description="COM_CONFIG_FIELD_SESSION_METADATA_GUEST_DESC"
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			label="COM_CONFIG_FIELD_SESSION_METADATA_GUEST_LABEL"
			description="COM_CONFIG_FIELD_SESSION_METADATA_GUEST_DESC"
			layout="joomla.form.field.radio.switcher"
			default="0"
```

**Comment by HLeithner** [comment]
Wrong approach in my opinion, we should try to create no user session if we have only a guest user. And only create the session if needed by a extension.

**Comment by brianteeman** [comment]
Good luck with that

**Comment by HLeithner** [comment]
> Good luck with that

I wrote already a poc a while ago

**Comment by Fedik** [comment]
@HLeithner I just extending existing feature :wink: 

> we should try to create no user session if we have only a guest user. And only create the session if needed by a extension.

That would be better of course, but that will be much to do, and I have no idea how.
And when someone will make it, it will be not hard to remove current "option".

**Comment by HLeithner** [comment]
true, other thing isn't select box with 3 options better?

**Comment by Fedik** [comment]
> true, other thing isn't select box with 3 options better?

I not sure.
Then probably I need to change the `session_metadata` value type to integer, that tehnically will be b/c. It is boolean currently.

**Comment by simbus82** [comment]
> This should improve **performance** for sites that does not use User registration/login feature.

I'm really interested in this aspect, performance. In case of many visitors then this solution can improve the performances avoiding the work due to the creation of the sessions ... correct?
What side effects, if any, will there be compared to the current CMS behavior for guests?

Thanks!

**Comment by HLeithner** [comment]
> I'm really interested in this aspect, performance. In case of many visitors then this solution can improve the performances avoiding the work due to the creation of the sessions ... correct? What side effects, if any, will there be compared to the current CMS behavior for guests?
> 
> Thanks!

sadly not, it only allow to disable metadata for guest users, you can disable metadata for all users if you like then no addition database query for updating the stats will be done. But still Joomla creates a Session for each user (request without session)

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on b2d406e0a1c888ed480271077b86408d0a489a95<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37459">issues.joomla.org/tracker/joomla-cms/37459</a>.</sub>

**Comment by joeforjoomla** [comment]
I have tested this item :red_circle: unsuccessfully on b2d406e0a1c888ed480271077b86408d0a489a95<br /><br />The new parameter is not b/c because it's set by default to 0 and it's disabled by default. This causes to lose metadata for guest users by default causing b/c issues to extensions relying on these data also for guest users.
The parameter must be '1' enabled by default.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37459">issues.joomla.org/tracker/joomla-cms/37459</a>.</sub>

**Comment by Fedik** [comment]
I have changed to enabled by default, test again

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 99e7bb90c40260c8d6109a089c53c907051232ae<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37459">issues.joomla.org/tracker/joomla-cms/37459</a>.</sub>

**Comment by joeforjoomla** [comment]
I have tested this item :white_check_mark: successfully on 99e7bb90c40260c8d6109a089c53c907051232ae<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37459">issues.joomla.org/tracker/joomla-cms/37459</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37459">issues.joomla.org/tracker/joomla-cms/37459</a>.</sub>

File: administrator/language/en-GB/com_config.ini [hunk 3]
```diff
@@ -167,6 +167,8 @@ COM_CONFIG_FIELD_SERVER_TIMEZONE_LABEL="Website Time Zone"
 COM_CONFIG_FIELD_SESSION_HANDLER_LABEL="Session Handler"
 COM_CONFIG_FIELD_SESSION_METADATA_DESC="If enabled, additional metadata about a user's session (including their username, user ID, and which application they are logged into) will be logged to the session database table.<br>If disabled, features dependent on this data will be unavailable."
 COM_CONFIG_FIELD_SESSION_METADATA_LABEL="Track Session Metadata"
+COM_CONFIG_FIELD_SESSION_METADATA_GUEST_DESC="If enabled, additional metadata about both registered and non registered user's session will be logged to the session database table. <br>If disabled, only data for registered user will be logged."
+COM_CONFIG_FIELD_SESSION_METADATA_GUEST_LABEL="Track Session Metadata for non registered Users"
```
Comment (Quy) [inline administrator/language/en-GB/com_config.ini]:
```suggestion
COM_CONFIG_FIELD_SESSION_METADATA_GUEST_LABEL="Track Session Metadata for Non Registered Users"
```

Comment (brianteeman) [reply administrator/language/en-GB/com_config.ini]: it was correct as it is. I would only use R if we are referring to the Registered group and not as a collective noun for all users

Comment (brianteeman) [reply administrator/language/en-GB/com_config.ini]: oops sorry I see that this is a label. 

Wouldnt it be better as "Track Guest Session Metadata"

Comment (Fedik) [reply administrator/language/en-GB/com_config.ini]: It was "extending" of previous label:
`COM_CONFIG_FIELD_SESSION_METADATA_LABEL="Track Session Metadata"`
to 
`COM_CONFIG_FIELD_SESSION_METADATA_GUEST_LABEL="Track Session Metadata for non registered Users"`

But can change to anything :)

Comment (roland-d) [reply administrator/language/en-GB/com_config.ini]: What should the language string be?

Comment (Fedik) [reply administrator/language/en-GB/com_config.ini]: I have changed to what @brianteeman sugested: "Track Guest Session Metadata"

**Pull request merged**